### PR TITLE
aggregator: add port support, unified for both modes

### PR DESCRIPTION
**Pull request by sttts**
Add a port field to the service reference of an APIService.

### CODE REVIEW

**Comment by deads2k** [comment]
@cheftako you want to take a look?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go [hunk 1]
```diff
@@ -96,7 +97,10 @@ func (c *APIServiceRegistrationController) sync(key string) error {
 		return err
 	}
 
-	c.apiHandlerManager.AddAPIService(apiService, c.getDestinationHost(apiService))
+	host := c.getDestinationHost(apiService)
+	if len(host) > 0 {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go]: If an APIService is updated to remove its referenced service (becomes local), we need to update to reflect that, right?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go [hunk 2]
```diff
@@ -110,16 +116,27 @@ func (c *APIServiceRegistrationController) getDestinationHost(apiService *apireg
...
+	case v1.ServiceTypeExternalName:
+		destinationHost = service.Spec.ExternalName
+		if len(destinationHost) == 0 {
+			return ""
+		}
+	}
+
+	port := int32(443)
+	if len(service.Spec.Ports) > 0 {
+		port = service.Spec.Ports[0].Port
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go]: I assumed that there is an order. If not, this does not make sense, totally agree. Actually, I do not want too many heuristic. I would prefer adding an explicit port name to the apiservice spec and only allow to leave that empty if there is exactly one port in a service. Wdyt?

**Comment by luxas** [comment]
@sttts Why not just support portName right away?

**Comment by sttts** [comment]
@deads2k @cheftako I rebased this onto the merged pod ip resolver code, reactivated our old resolver tests and unified both resolvers. I would very much like to get this into 1.7 before people start using the new api, especially the requirements around service ports and the service name syntax `scheme:name:port`.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go [hunk 3]
```diff
@@ -29,51 +29,103 @@ import (
...
-		return nil, errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+		return "", "", "", errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+	}
+
+	if len(svcScheme) == 0 {
+		svcScheme = "https"
+	}
+	if len(portStr) == 0 {
+		switch svcScheme {
+		case "http":
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: You can't safely proxy over http.  I think we should always force https.

**Comment by deads2k** [comment]
> especially the requirements around service ports and the service name syntax scheme:name:port.

We have to use SSL in order to identify the aggregator from the user API server.  For ports, my first choice is always 443 on the service (the endpoint thing is going to have to do the translation), my second choice is to add a port field and the endpoint thing still has to do the translation from service to endpoint port.

File: staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go]: My argument for http was that it's useful during development without certs. But if we can ignore self-signed certs anyway, there is not much need for http.

**Comment by sttts** [comment]
I have extracted the uncontroversial part of the code without the scheme+port syntax into https://github.com/kubernetes/kubernetes/pull/46623. ptal.

**Comment by sttts** [comment]
> my second choice is to add a port field and the endpoint thing still has to do the translation from service to endpoint port.

I favor this one. Let us add a `Port *IntString` field which defaults to 443.

**Comment by sttts** [comment]
@deads2k @cheftako added the port field to ServiceReference. ptal

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-verify test this

**Comment by sttts** [comment]
We decided not to implement custom port support. So closing it for now.

**Comment by WIZARD-CXY** [comment]
@sttts I wonder why " we decided not to implement custom port support"? I think it is nice to have.

**Comment by sttts** [comment]
@WIZARD-CXY because it is only "nice to have", i.e. you can easily use a default port.

**Pull request closed without merging**